### ISSUE TITLE
Async Chunks: Freezing server, Sync Chunks: Empty chunks

### ISSUE THREAD

**Issue by dogboy21** [PaperMC/Paper#1586]
**No plugins installed**
**Paper Version:** git-Paper-[dcc764db](https://github.com/PaperMC/Paper/commit/dcc764db2195381cfe2633d5e0ad7ff806b5b3ea)

After updating our servers from 1.12.2 to 1.13.1 I've noticed some weird behaviour with chunk loading / rendering in general.

When `async-chunks` is enabled in the `paper.yml` config, the server freezes immediately after a player joins the server and eventually crashes.  -> [Stacktrace](https://gist.github.com/dogboy21/ce38e6f62845d81e00c7d742d24a5ed1)

After disabling async chunks, the server freeze/crash was fixed, but another issue came up: Some chunks didn't load at all on the client, others only showed up after spamming F3+A several times. 
I've already seen #1504 which seems to be related to this issue, however it's still present in the newest version. The server spams [this](https://gist.github.com/dogboy21/188e811bc4538472fac71126a5330066) all over the console when the player flies towards missing chunks.
Chunks that are completely missing (F3+A not helping) are very strange in their behaviour: a player can place blocks in them and collission also applies as if the blocks were present, but they only show up after a complete server restart, even relogging doesn't help. 
At least it did in my initial testing. During stacktrace collection for this issue I tried it again, this time however the server crashed immediately after joining after the restart. -> [Stacktrace](https://gist.github.com/dogboy21/6940ff7f64bfb541ea1ce7945710d600)

Additionally, it seems that some blocks aren't correctly send to the client:
![Missing blocks](https://ss.myftb.de/buLunLbq.png)
Between the fence and the hopper **should** be a cobble wall. Some of them showed up after a server restart, others didn't

**Comment by aikar** [comment]
Can you paste full log? Missing lots of important parts.

**Comment by dogboy21** [comment]
Log with async chunks enabled: https://gist.github.com/dogboy21/1e50a5eaa7d57f97d7d2ac36394c5076
Log after disabling async chunks: https://gist.github.com/dogboy21/ec6ded25398ab0acf204a2acd729c081

**Comment by aikar** [comment]
please let it fully crash. You're missing the full stack trace at the end is what I'm trying to see.

**Comment by dogboy21** [comment]
https://gist.github.com/dogboy21/4598bf9a9321b8bd1f68ed9d9a3ad9e1

**Comment by aikar** [comment]
Ok I got the important part from that
```
[18:58:33] [Paper Watchdog Thread/ERROR]: Current Thread: PaperChunkLoader-4
[18:58:33] [Paper Watchdog Thread/ERROR]: 	PID: 42 | Suspended: false | Native: false | State: BLOCKED
[18:58:33] [Paper Watchdog Thread/ERROR]: 	Stack:
[18:58:33] [Paper Watchdog Thread/ERROR]: 		it.unimi.dsi.fastutil.longs.Long2ObjectFunctions$SynchronizedFunction.containsKey(Long2ObjectFunctions.java:221)
[18:58:33] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.v1_13_R2.ChunkProviderServer.isLoaded(ChunkProviderServer.java:503)
[18:58:33] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.v1_13_R2.ChunkRegionLoader.check(ChunkRegionLoader.java:124)
```
Weird issue 1 here is this issue I see quite a few people have where the 1.12 World Conversion code is triggering some blocks to 'pop' and set to air, causing a chain reaction of Physics Events to pop chains.

But this heavy code is running on main holding a lock, and then stuff async is trying to check main.

We need to move these post processing tasks to happen outside of the chunk map so it's not synchronized.
That stops the crash. I don't know why things are popping in first place though.

**Comment by aikar** [comment]
Fixed in 0d1ca37, build 378+
